syncing/src/test/kotlin/maru/sync/pipeline/BeaconChainDownloadPipelineFactoryTest.kt

```diff
@@ -0,0 +1,225 @@
+    val completionFuture = pipeline.start(executorService)
+
+    // Wait for completion
+    completionFuture.get(5, TimeUnit.SECONDS)
+
+    // Verify all blocks were imported
+    verify(blockImporter, atLeastOnce()).importBlock(any())
+  }
+
+  @Test
+  fun `pipeline handles single block range`() {
+    val peer = mock<MaruPeer>()
+    whenever(peerLookup.getPeers()).thenReturn(listOf(peer))
+
+    val blocks = listOf(DataGenerators.randomSealedBeaconBlock(42uL))
+    val response = mock<BeaconBlocksByRangeResponse>()
+    whenever(response.blocks).thenReturn(blocks)
+    whenever(peer.sendBeaconBlocksByRange(42uL, 1uL)).thenReturn(SafeFuture.completedFuture(response))
+
+    whenever(blockImporter.importBlock(any())).thenReturn(
+      SafeFuture.completedFuture(ValidationResult.Companion.Valid),
+    )
+
+    val pipeline = factory.createPipeline(42uL, 42uL)
+    val completionFuture = pipeline.start(executorService)
+
+    completionFuture.get(5, TimeUnit.SECONDS)
+
+    verify(peer).sendBeaconBlocksByRange(42uL, 1uL)
+    verify(blockImporter).importBlock(blocks[0])
+  }
+
+  @Test
+  fun `pipeline with large request size processes correct ranges`() {
+    val largeRequestSizeFactory =
+      BeaconChainDownloadPipelineFactory(
+        blockImporter = blockImporter,
+        metricsSystem = NoOpMetricsSystem(),
+        peerLookup = peerLookup,
+        downloaderParallelism = 1,
+        requestSize = 100u,
+      )
+
+    val peer = mock<MaruPeer>()
+    whenever(peerLookup.getPeers()).thenReturn(listOf(peer))
+
+    // Create blocks for range [0, 50]
+    val blocks = (0uL..50uL).map { DataGenerators.randomSealedBeaconBlock(it) }
+    val response = mock<BeaconBlocksByRangeResponse>()
+    whenever(response.blocks).thenReturn(blocks)
+    whenever(peer.sendBeaconBlocksByRange(0uL, 51uL)).thenReturn(SafeFuture.completedFuture(response))
+
+    whenever(blockImporter.importBlock(any())).thenReturn(
+      SafeFuture.completedFuture(ValidationResult.Companion.Valid),
+    )
+
+    val pipeline = largeRequestSizeFactory.createPipeline(0uL, 50uL)
+    val completionFuture = pipeline.start(executorService)
+
+    completionFuture.get(5, TimeUnit.SECONDS)
+
+    // Should make only one request since request size (100) is larger than range
+    verify(peer).sendBeaconBlocksByRange(0uL, 51uL)
+  }
+
+  @Test
+  fun `pipeline handles empty block response`() {
+    val peer = mock<MaruPeer>()
+    whenever(peerLookup.getPeers()).thenReturn(listOf(peer))
+
+    val response = mock<BeaconBlocksByRangeResponse>()
+    whenever(response.blocks).thenReturn(emptyList())
+    whenever(peer.sendBeaconBlocksByRange(any(), any())).thenReturn(SafeFuture.completedFuture(response))
+
+    val pipeline = factory.createPipeline(100uL, 110uL)
+    val completionFuture = pipeline.start(executorService)
+
+    completionFuture.get(5, TimeUnit.SECONDS)
+
+    // Pipeline should complete without errors
+    assertThat(completionFuture).isCompleted
```

Comment:
Shouldn't we add more assertions that block import was called for all the expected blocks?